### ISSUE TITLE
Add an `@exported_previously_as()` annotation to avoid losing data

### ISSUE THREAD

**Issue by AdriaandeJongh** [godotengine/godot-proposals#10644]
### Describe the project you are working on

[Rift Riff](https://store.steampowered.com/app/2800900/Rift_Riff/) and soon a sequel to [Hidden Folks](https://hiddenfolks.com).

### Describe the problem or limitation you are having in your project

Cleaning up and refactoring a script by renaming exported variables means you also have to re-set the variables in the editor. Unity has the `[FormerlySerializedAs("")]` C# attribute which is a huge time saver. Godot should have something similar, or even better!

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

An `@exported_previously_as("<variable name>")` annotation makes Godot look at the serialized data and put any data it finds for the variable name passed to the annotation into the exported variable this annotation applies to.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Let's say I have this variable which is set up in the editor:

```
@export var list_of_things: Array[Node2D]
```

When I then refactor it to this:

```
@exported_previously_as("list_of_things")
@export var a_better_named_list: Array[Node2D]
```

Any exported data that was previously in `list_of_things` is now set up in `a_better_named_list`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

One not-so-easy workaround I have found is to add a new export variable to a script without overwriting the old data and then write an editor tool that finds all applicable objects and manually places the data of the old exported variable into the new.

### Is there a reason why this should be core and not an add-on in the asset library?

Annotations feel like they should be core, al though I don't even know whether annotations can be added through addons? That would be cool.

**Comment by AdriaandeJongh** [comment]
Ah, yeah, a clever workaround indeed.

**Comment by AdriaandeJongh** [comment]
I'm seeing thumbs down and confused reactions to my proposal. Perhaps someone can explain their reasoning behind those 'downvotes'?

**Comment by AThousandShips** [comment]
I'd say if people had direct opinions and feedback to share they would, otherwise it's just a matter of expressing opinion

**Comment by AThousandShips** [comment]
I think the solution to this is to add a section in the documentation for how to add legacy support and adding boilerplate code to `_get` and `_set` to handle old properties like this

My personal philosophy and preference for GDScript is to have as little hidden code and complex syntactic sugar as possible, annotations like `@export` are primarily external to the code and just deals with registering the export data, `@onready` does add code to the initialization step, but otherwise GDScript has very little hidden code, it's largely what you see is what you get, and there isn't much in the way of "this annotation adds a whole block of code hidden somewhere"

So I think this should be documented as a *workflow* rather than adding annotations with magic code running in the background, and if we add some form of macro or boilerplate feature support down the line that would be useful to add as features to your macro library for a standard code block to handle the legacy feature

**Comment by AdriaandeJongh** [comment]
I think I totally agree with you on those points, and would even love to contribute to the docs to add the workflow you suggest. Where in the docs do you think this fits best?

**Comment by AThousandShips** [comment]
I'd suggest adding it under [GDScript exported properties](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_exports.html)

**Comment by AThousandShips** [comment]
Why would you need `_get`? The `_set` part is for when a file is loaded and the property in the file is assigned from the file, which would use the `_set` method

**Comment by AdriaandeJongh** [comment]
In my tests, both in a script with the `@tool` annotation and without, `_set()` is never called after renaming an `@export`'ed variable.

**Comment by AThousandShips** [comment]
Did you have the scene open while renaming?

**Comment by AdriaandeJongh** [comment]
Ahh, right. That's the thing I was missing in this process. The scene would have to be closed, then the variable renamed, and then reopened for this to work. I'll admit this is very finicky...

**Comment by AThousandShips** [comment]
I don't see how adding an annotation would work around that, don't know how it would handle that rename step

But indeed the code with `_set` works perfectly well with a `@tool` script

**Comment by AdriaandeJongh** [comment]
Alright – now that the process is clear to me I'll make a PR for the docs. Thanks for your help.

**Comment by eon-s** [comment]
This is a much needed feature for sane refactoring, the most common case when working in teams particularly is that a script can change a variable name and could be used in a scene that then, at export time will fail, on opening time will fail, etc.

The workaround is too risky because the tool system needs modifications all over the scripts to prevent running game logic in the editor, so is not desirable, also it means that every single script will need to be a tool script to survive refactors in the whole project lifetime.

**Comment by AdriaandeJongh** [comment]
I agree adding `@tool` to scripts is not desirable. @dalexeev also pointed out in my PR to add the workflow discussed above that someone can also A) open the entire project in a text editor and find-and-replace the property you want to rename or B) use a mix of @export_storage on old properties (without removing them) and create new properties with getters / setters so that the values are linked. All good workarounds, yes, but also risky, lot's of work, and undesirable.

**Comment by molingyu** [comment]
Rather than adding a new annotation, it is better to add a new `old_names` parameter to the export annotation.

```
@export(old_names=[ "list_of_thingslist_of_things" ]) var a_better_named_list: Array[Node2D]
```

**Comment by och-och** [comment]
I think the proposed solution while mitigating the loss of data would just add clutter in the script and not actually help with fixing outdated resource files, as I assume those don't get updated to have the new property name? And if they do, (I, again, assume) this will only affect resources when they're opened in the editor, and that leaves you with outdated resource sitting in the project; that's not cool, I think.
A better solution in my opinion is for the engine to not discard the unbound data when opening a scene or resource, but instead to prompt the user with a popup or better with a warning icon in the inspector (and the node tree next to the affected nodes) that opens up a dialog that allows replacing the old property name with a new one which could be selected from a list of exported properties that have a compatible type. And this will update all resources and nodes that have this script (or its derivative) attached.
This will provide an actual solution to the problem without having to keep all the old property names inside the script.
🧙‍♂️🦖🐸🤔⚠️⚠️⚠️

**Comment by eon-s** [comment]
The thing is that the engine/editor shouldn't assume what the user wanted to do, these actions should be explicit otherwise common actions like changing a script will end in a lot of unused bloat in scenes and resources (also hard to maintain because orphan legacy data will keep piling on resources), that's why explicit intentions (editor, please, migrate this old name to this new one) are usually better.

**Comment by Creta5164** [comment]
Just discovered about this proposal, may related to https://github.com/godotengine/godot-proposals/issues/5945.
But seems @AThousandShips's answer (using `_set` to migrate them) seems very clean way to do it, I gonna try about this.

**Comment by Bifi1992** [comment]
It seems both @AdriaandeJongh and @AThousandShips agree that the provided workaround of using a setter and a tool script is very finicky and prone to errors from what I've read.

So, I agree with @och-och wish for
> A better solution in my opinion is for the engine to not discard the unbound data when opening a scene or resource

I think that the solutions to the issue have not been explored, yet. So here's the beginning of a list of mentioned options:

* @och-och wants an editor prompt on `@export` renaming
  * cons:
    * by @eon-s - https://github.com/godotengine/godot-proposals/issues/10644#issuecomment-2665921287
* @fkeyzuwu suggested in https://github.com/godotengine/godot-proposals/issues/13186 to provide a refactoring function from the editor (this has sadly instantly been closed with reference to this)
* I saw someone suggesting an additional new export annotation (gladly remind me of the person/thread, I will update this post)
* I was thinking about extending the core `@export` annotation with an inspector UID that can optionally be provided
  * cons:
    * what if the type changes?

Maybe you guys are up to a discussion of potential solutions for this issue instead of relying on the current workaround =)

**Comment by AThousandShips** [comment]
I did not say it was very finicky or error prone so no that's not something I agree on

**Comment by Bifi1992** [comment]
> I did not say it was very finicky or error prone so no that's not something I agree on

see https://github.com/godotengine/godot-proposals/issues/10644#issuecomment-2329134329 

> So yes, it's a bit finicky

PS: I don't really wanna argue, though 🙇

**Comment by AThousandShips** [comment]
I said it was "a bit finicky", so that's my opinion on that, not "very finicky prone to errors", no need for any argument, I just don't agree with what you said I agreed with and let's leave it at that thank you!

**Comment by fkeyzuwu** [comment]
> [@fkeyzuwu](https://github.com/fkeyzuwu) suggested in [Add option to refactor `@export` variable name without resetting values #13186](https://github.com/godotengine/godot-proposals/issues/13186) to provide a refactoring function from the editor (this has sadly instantly been closed with reference to this)

1. If this would be implemented, i'm assuming it would have to be after https://github.com/godotengine/godot/pull/102380. Since this would use whatever this API would give us + some extra.
2. Even if this would be done, this would only be a feature inside the godot script editor, not in any external IDEs. So I'm not sure how much I like it. Basically because this has to do with scenes and not *just* code, it might be better to be done in a way which no matter what IDE you use, it will always work.

**Comment by BlueberryGecko** [comment]
Just because I haven't seen it yet, I want to add that a simple, but incomplete, solution is to

1) rename the property in the script (e.g. `timerMax` -> `triggerTimerMax`)
2) save everything
3) open the `tscn` and `tres` files in a text editor and text replace the `timerMax = 0.5` with `triggerTimerMax = 0.5`.

This works because Godot (4.5) apparently keeps the `timerMax = 0.5` lines in the PackedScene file even after I've renamed the script property.

It's obviously less than ideal, but if you're comfortable with regex and sed and the like, this at least makes it possible to rename a variable with, say, 30 seconds of effort, which is still better than infinite seconds (as in "I don't trust this rename, so I'll never do it") of effort ;)
